### ISSUE TITLE
Training loop may not encourage the imputer to infer missing data ?

### ISSUE THREAD

**Issue by vilhess** [jicongfan/ImAD-Anomaly-Detection-With-Missing-Data#3]
Hi,

Thanks for the very interesting paper and for open-sourcing the code.

I have a question regarding the implementation of the training loop.

The goal of the imputer is to infer missing values. However, in both the normal and abnormal settings, the loss is computed **only on visible samples**, not on the masked ones. This seems to encourage the model to learn an identity mapping rather than to impute missing values.

For example, in the normal setting:

```python
# imputation loss (normal setting)
imputed_loss = torch.sum(torch.mul((imputed_data  - inputs), masks) ** 2, dim=tuple(range(1, inputs.dim()))) * self.beta
```

Here, `mask` is 1 for visible inputs and 0 otherwise.
The abnormal setting appears to use the same logic.

My intuition is that the projector may help the model generate values that do not disrupt the latent space and potentially depending on the other variables, but this still does not seem to train the model to infer missing values.

Am I misunderstanding something, or is there a potential issue with the implementation?

Thanks in advance for your help!

**Comment by xiaofeng-github** [comment]
Yes, the imputer designed only focuses on the visible values since the goal of this work is to identify the anomalies in incomplete data.  In a real scenario that the original data are incomplete, there is no any prior information for the missing values. The masks used aim to mimic the different missing mechanisms. In other word, you cannot utilize any information about the missing values in this setting.
